storm-server/src/main/java/org/apache/storm/daemon/supervisor/Container.java

```diff
@@ -290,49 +301,368 @@ private boolean isWindowsProcessAlive(long pid, String user) throws IOException
+        }
+        File f = new File(fpath);
+        if (!f.exists()) {
+            LOG.error("Cannot determine owner of non-existent file {}", fpath);
+            return -1;
+        }
+        LOG.debug("CMD: ls -n {}", fpath);
+        ProcessBuilder pb = new ProcessBuilder("ls", "-n", fpath);
+        pb.redirectError(ProcessBuilder.Redirect.INHERIT);
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(pb.start().getInputStream()))) {
+            String line = in.readLine();
+            LOG.debug("CMD-OUTLINE: {}", line);
+            line = line.trim();
+            String[] parts = line.split("\\s+");
+            if (parts.length < 3) {
+                LOG.error("Expecting at least 3 space separated fields in \"ls -n {}\" output, got {}", fpath, line);
+                return -1;
+            }
+            try {
+                return Integer.parseInt(parts[2]);
+            } catch (NumberFormatException ex) {
+                LOG.error("Expecting at third field {} to be numeric UID \"ls -n {}\" output, got {}", parts[2], fpath, line);
+                return -1;
+            }
+        } catch (IOException ex) {
+            LOG.error(String.format("Cannot read output of command \"ls -n %s\"", fpath), ex);
+            return -1;
+        }
+    }
+
+    /**
+     * Get UID of the owner to the workerId PIDs Root directory.
+     *
+     * @return User ID (UID) of owner of the PID file, -1 if file is missing.
+     */
+    private int getPidPathOwnerUid(String workerId, long pid) {
+        return getPathOwnerUId(ConfigUtils.workerPidPath(conf, workerId, pid));
+    }
+
+    /**
+     * Find if all processed for the user on workId are dead.
+     * This method attempts to optimize the calls by:
+     * <p>
+     *     <li>creating a collection of ProcessIds and checking all of them at once</li>
+     *     <li>using userId one Posix systems instead of user</li>
+     * </p>
+     *
+     * @return true if all processes for the user are dead on the worker
+     * @throws IOException if external commands have exception.
+     */
     @Override
     public boolean areAllProcessesDead() throws IOException {
         Set<Long> pids = getAllPids();
         String user = getRunWorkerAsUser();
 
         boolean allDead = true;
-        for (Long pid : pids) {
-            LOG.debug("Checking if pid {} owner {} is alive", pid, user);
-            if (!isProcessAlive(pid, user)) {
-                LOG.debug("{}: PID {} is dead", workerId, pid);
+        try {
+            if (pids.isEmpty()) {
+                return true;
+            }
+            if (ServerUtils.IS_ON_WINDOWS) {
+                return allDead = !isAnyProcessAlive(pids, user);
+            }
+            // optimized for Posix - try to use uid
+            if (!cachedUserToUidMap.containsKey(user)) {
+                int uid = getPidPathOwnerUid(workerId, pids.iterator().next());
```

Comment:
We don't have to check the owner of pid file. Checking the owner of the worker-id directory should be good enough.

Reply:
Replaced with check on directory instead of file. also changed corresponding command to use "ls -dn" instead of "ls -n".